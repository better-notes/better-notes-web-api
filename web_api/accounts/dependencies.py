import secrets
from typing import Callable
from aioredis import create_redis_pool
from aioredis.commands import Redis
from fastapi import Depends
from motor import motor_asyncio  # type: ignore
from passlib.hash import bcrypt  # type: ignore
from web_api.accounts import repositories, usecases
from web_api.commons.dependencies import get_mongo_client
from web_api.settings import Settings


def get_user_repository(
    client: motor_asyncio.AsyncIOMotorClient = Depends(get_mongo_client),
    settings: Settings = Depends(lambda: Settings()),
) -> repositories.UserRepository:
    return repositories.UserRepository(client=client, settings=settings)


def get_bcrypt() -> bcrypt:
    return bcrypt


def get_account_register_use_case(
    user_repository: repositories.UserRepository = Depends(
        get_user_repository
    ),
    hasher: bcrypt = Depends(get_bcrypt),
) -> usecases.AccountRegisterUseCase:
    return usecases.AccountRegisterUseCase(
        user_repository=user_repository, hasher=hasher
    )


def get_account_authenticate_use_case(
    user_repository: repositories.UserRepository = Depends(
        get_user_repository
    ),
    hash_verifier: bcrypt = Depends(get_bcrypt),
) -> usecases.AccountAuthenticateUseCase:
    return usecases.AccountAuthenticateUseCase(
        user_repository=user_repository, hash_verifier=hash_verifier
    )


async def get_redis(settings: Settings = Depends(lambda: Settings())) -> Redis:
    return await create_redis_pool(settings.REDIS_ADDRESS)


def get_user_session_repository(
    client: Redis = Depends(get_redis),
) -> repositories.UserSessionRepository:
    return repositories.UserSessionRepository(client=client)


def get_account_session_id_generator() -> Callable[[], str]:
    return secrets.token_hex


def get_account_session_interactor(
    user_session_repository: repositories.UserSessionRepository = Depends(
        get_user_session_repository
    ),
    user_session_id_generator: Callable[[], str] = Depends(
        get_account_session_id_generator
    ),
) -> usecases.AccountSessionInteractor:
    return usecases.AccountSessionInteractor(
        user_session_repository=user_session_repository,
        generate_user_session_id=user_session_id_generator,
    )
