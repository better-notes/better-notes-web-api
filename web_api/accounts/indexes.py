from motor import motor_asyncio


async def create_username_unique_index(
    users_collection: motor_asyncio.AsyncIOMotorCollection,
) -> None:
    """Unique index on account username."""
    await users_collection.create_index('username', unique=True)
