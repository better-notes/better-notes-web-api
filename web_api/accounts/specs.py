import dataclasses

from web_api.commons.specs import Specification


@dataclasses.dataclass
class GetUserByUsernameSpecification(Specification):
    username: str

    def get_query(self) -> dict[str, str]:
        return {'username': self.username}
