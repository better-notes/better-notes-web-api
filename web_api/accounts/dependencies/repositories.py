import aioredis
from aioredis.commands import Redis
from fastapi import Depends
from motor import motor_asyncio

from web_api.accounts import repositories
from web_api.commons.dependencies import get_mongo_client, get_settings
from web_api.settings import Settings


def get_account_repository(
    client: motor_asyncio.AsyncIOMotorClient = Depends(get_mongo_client),
    settings: Settings = Depends(get_settings),
) -> repositories.AccountRepository:
    """Get account repository."""
    return repositories.AccountRepository(client=client, settings=settings)


async def get_redis(settings: Settings = Depends(get_settings)) -> Redis:
    """Get redis client."""
    pool = await aioredis.create_redis_pool(settings.redis_address)
    yield pool
    pool.close()
    await pool.wait_closed()


def get_account_session_repository(
    client: Redis = Depends(get_redis),
) -> repositories.AccountSessionRepository:
    """Get account session repository."""
    return repositories.AccountSessionRepository(client=client)
