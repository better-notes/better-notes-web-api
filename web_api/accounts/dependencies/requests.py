from fastapi.param_functions import Cookie

from web_api.accounts.values import AuthenticationTokenValue


def get_authentication_token_value(
    authentication_token: str = Cookie(...),
) -> AuthenticationTokenValue:
    """Get authentication token value from cookie."""
    return AuthenticationTokenValue(value=authentication_token)
