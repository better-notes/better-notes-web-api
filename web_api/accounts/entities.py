from datetime import datetime
from web_api.accounts import values
from web_api.commons import entities


class UserEntity(entities.Entity):
    id_: str
    username: str
    password_hash: str
    created_at: datetime


class UserSessionEntity(entities.Entity):
    token: values.AuthenticationToken
    user: UserEntity
