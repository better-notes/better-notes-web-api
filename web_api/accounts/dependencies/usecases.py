import secrets
from typing import Callable

from fastapi import Depends
from passlib.hash import bcrypt  # type: ignore

from web_api.accounts import repositories, usecases
from web_api.accounts.dependencies.repositories import (
    get_account_repository,
    get_account_session_repository,
)


def get_bcrypt() -> bcrypt:
    """Get bcrypt hasher."""
    return bcrypt


def get_account_register_use_case(
    user_repository: repositories.AccountRepository = Depends(get_account_repository),
    hasher: bcrypt = Depends(get_bcrypt),
) -> usecases.AccountRegisterUseCase:
    """Get account registration use case."""
    return usecases.AccountRegisterUseCase(user_repository=user_repository, hasher=hasher)


def get_account_authenticate_use_case(
    user_repository: repositories.AccountRepository = Depends(get_account_repository),
    hash_verifier: bcrypt = Depends(get_bcrypt),
) -> usecases.AccountAuthenticateUseCase:
    """Get account authentication use case."""
    return usecases.AccountAuthenticateUseCase(
        user_repository=user_repository, hash_verifier=hash_verifier,
    )


def get_account_session_id_generator() -> Callable[[], str]:
    """Get account session id generator."""
    return secrets.token_hex


def get_account_session_interactor(
    user_session_repository: repositories.AccountSessionRepository = Depends(
        get_account_session_repository,
    ),
    user_session_id_generator: Callable[[], str] = Depends(get_account_session_id_generator),
) -> usecases.AccountSessionInteractor:
    """Get account session interactor."""
    return usecases.AccountSessionInteractor(
        account_session_repository=user_session_repository,
        generate_user_session_id=user_session_id_generator,
    )
