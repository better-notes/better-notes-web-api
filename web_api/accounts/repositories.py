import dataclasses
from datetime import datetime, timedelta, timezone

import bson
from aioredis.commands import Redis
from motor import motor_asyncio
from pymongo.results import InsertOneResult

from web_api import commons
from web_api.accounts import entities, values
from web_api.commons.repositories import AbstractRepository
from web_api.commons.values import Paging
from web_api.settings import Settings


@dataclasses.dataclass
class AccountRepository(AbstractRepository):
    """Repository for accounts."""

    client: motor_asyncio.AsyncIOMotorClient
    settings: Settings

    def __post_init__(self) -> None:
        db = self.client[self.settings.MONGO_DATABASE]
        self.users_collection = db['users']

    async def add(
        self, *, account_value_list: list[values.AccountValue],
    ) -> list[entities.AccountEntity]:
        inserted_entities = []

        for account_value in account_value_list:
            value_data = account_value.dict()

            created_at = datetime.now(timezone.utc)
            user_insert: InsertOneResult = (
                await self.users_collection.insert_one(
                    {**value_data, 'created_at': created_at}
                )
            )

            value_data.pop('password_hash')
            inserted_entities.append(
                entities.AccountEntity(
                    id_=str(user_insert.inserted_id),
                    created_at=created_at,
                    **value_data,
                )
            )

        return inserted_entities

    async def get(
        self, *, spec: commons.specs.Specification, paging: Paging,
    ) -> list[entities.AccountEntity]:
        cursor = (
            self.users_collection.find(spec.get_query())
            .limit(paging.limit)  # noqa: WPS348 TODO: discuss
            .skip(paging.offset)  # noqa: WPS348 TODO: discuss
        )

        account_entity_list = []
        async for account_entity_dict in cursor:
            account_entity_dict.pop('password_hash')
            account_entity_list.append(
                entities.AccountEntity(
                    id_=str(account_entity_dict.pop('_id')),
                    **account_entity_dict,
                ),
            )
        return account_entity_list

    async def update(
        self, *, account_entity_list: list[entities.AccountEntity],
    ) -> list[entities.AccountEntity]:
        for entity in account_entity_list:
            entity_data = entity.dict(exclude={'id_'})
            await self.users_collection.update_one(
                {'_id': bson.ObjectId(entity.id_)}, {'$set': entity_data},
            )
        return account_entity_list

    async def delete(
        self, *, account_entity_list: list[entities.AccountEntity],
    ) -> list[entities.AccountEntity]:
        id_value_list = []
        for entity in account_entity_list:
            id_value_list.append(bson.ObjectId(entity.id_))

        await self.users_collection.delete_many(
            {'_id': {'$in': id_value_list}},
        )
        return account_entity_list

    async def get_password_hash(
        self, *, account_entity: entities.AccountEntity,
    ) -> str:
        """Get password hash for given account entity."""
        account_entity_projection = await self.users_collection.find_one(
            {'_id': bson.ObjectId(account_entity.id_)}, ['password_hash'],
        )
        return account_entity_projection['password_hash']


@dataclasses.dataclass
class AccountSessionRepository(AbstractRepository):
    """Repository for account sessions."""

    client: Redis

    async def add(
        self,
        *,
        account_session_entity_list: list[entities.AccountSessionEntity],
    ) -> list[entities.AccountSessionEntity]:
        for account_session_entity in account_session_entity_list:
            await self.client.set(
                account_session_entity.token.value,
                account_session_entity.json(),
                expire=timedelta(days=7).seconds,
            )

        return account_session_entity_list

    async def get(
        self,
        *,
        authentication_token_value_list: list[values.AuthenticationTokenValue],
    ) -> list[entities.AccountSessionEntity]:
        account_session_entity_list = []

        for authentication_token_value in authentication_token_value_list:
            entity_json = await self.client.get(
                authentication_token_value.value,
            )
            account_session_entity_list.append(
                entities.AccountSessionEntity.parse_raw(entity_json),
            )

        return account_session_entity_list
