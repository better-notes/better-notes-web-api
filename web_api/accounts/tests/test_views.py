from syrupy.filters import props
from web_api.accounts.tests.factories import (
    AccountRegisterUseCaseFactory,
    AuthenticationCredentialsFactory,
    RegistrationCredentialsFactory,
)
from web_api.accounts.dependencies import get_user_session_id_generator


async def test_register(client, app, reverse_route, snapshot):
    app.dependency_overrides[
        get_user_session_id_generator
    ] = lambda: lambda: 'session_id'

    registration_credentials = RegistrationCredentialsFactory()
    response = await client.post(
        reverse_route('register'), json=registration_credentials.dict()
    )

    assert response.json() == snapshot(exclude=props('id_', 'created_at'))


async def test_authenticate(client, app, reverse_route, snapshot):
    app.dependency_overrides[
        get_user_session_id_generator
    ] = lambda: lambda: 'session_id'

    password = 'test_password'
    registration_credentials = RegistrationCredentialsFactory(
        password1=password, password2=password
    )
    account_register_use_case = AccountRegisterUseCaseFactory()
    user_entity = await account_register_use_case.register(
        registration_credentials=registration_credentials
    )

    authentication_credentials = AuthenticationCredentialsFactory(
        username=user_entity.username, password=password
    )
    response = await client.post(
        reverse_route('authenticate'), json=authentication_credentials.dict()
    )

    assert response.json() == snapshot(exclude=props('id_', 'created_at'))
