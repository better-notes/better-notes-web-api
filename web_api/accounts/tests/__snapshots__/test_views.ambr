# name: test_authenticate
  <class 'dict'> {
    'value': 'session_id',
  }
---
# name: test_register
  <class 'dict'> {
    'value': 'session_id',
  }
---
